cdc/generation.cc

```diff
@@ -173,118 +167,34 @@ topology_description generate_topology_description(
-                if (s.is_set()) {
-                    some_stream = s;
-                    break;
-                }
-            }
-            if (some_stream.is_set()) {
-                break;
-            }
-        }
-
-        assert(idx != entries.size() && some_stream.is_set());
-
-        // Iterate over all ranges in the clockwise direction, starting with the one we found a stream for.
-        for (size_t off = 0; off < entries.size(); ++off) {
-            auto& ss = entries[(idx + off) % entries.size()].streams;
-
-            int last_set_stream_idx = ss.size() - 1;
-            while (last_set_stream_idx > -1 && !ss[last_set_stream_idx].is_set()) {
-                --last_set_stream_idx;
-            }
-
-            if (last_set_stream_idx == -1) {
+        static thread_local std::mt19937_64 rand_gen(std::random_device().operator()());
+        static thread_local std::uniform_int_distribution<int64_t> rand_dist(std::numeric_limits<int64_t>::min());
+        dht::sharder sharder(entry.streams.size(), entry.sharding_ignore_msb);
+        int64_t next_token_for_unallocated_shard = dht::token::to_int64(prev_token) + 1;
+        int64_t upper_bound = dht::token::to_int64(entry.token_range_end);
+        for (size_t shard_idx = 0; shard_idx < entry.streams.size(); ++shard_idx) {
+            int64_t token;
+            auto t = sharder.token_for_next_shard(prev_token, shard_idx);
+            if (t > entry.token_range_end) {
                 cdc_log.warn(
-                        "CDC wasn't able to generate any stream for vnode ({}, {}]. We'll use another vnode's streams"
-                        " instead. This might lead to inconsistencies.",
-                        tokens[(idx + off + entries.size() - 1) % entries.size()], tokens[(idx + off) % entries.size()]);
-
-                ss[0] = some_stream;
-                last_set_stream_idx = 0;
-            }
-
-            some_stream = ss[last_set_stream_idx];
-
-            // Replace 'unset' stream ids with indexes below last_set_stream_idx
-            for (int s_idx = last_set_stream_idx - 1; s_idx > -1; --s_idx) {
-                if (ss[s_idx].is_set()) {
-                    some_stream = ss[s_idx];
-                } else {
-                    ss[s_idx] = some_stream;
-                }
-            }
-            // Replace 'unset' stream ids with indexes above last_set_stream_idx
-            for (int s_idx = ss.size() - 1; s_idx > last_set_stream_idx; --s_idx) {
-                if (ss[s_idx].is_set()) {
-                    some_stream = ss[s_idx];
+                        "CDC wasn't able to generate any stream on shard {} on vnode ({}, {}]. "
+                        "This shard will have its stream located on another shard in this vnode instead. "
+                        "This might lead to worse performance.",
+                        shard_idx, prev_token, entry.token_range_end);
+                token = next_token_for_unallocated_shard;
```

Comment:
How is it possible that we'll fail here?

Reply:
This happens when a vnode is small and it does not have tokens for each shard.

Reply:
Then it doesn't need a stream for each shard. Failure is impossible.

Reply:
I guess you're right because the Base Table won't have any data belonging to such shards as well. Will change this. Thanks.

Reply:
done.